### PR TITLE
Add Torch layers support

### PR DESCRIPTION
**Pull request by trbromley**
This PR provides a TorchLayer class for converting a PennyLane QNode to a Torch layer, which can then be used with standard `torch.nn` features such as the Module/Sequential models for constructing neural networks.

Note that a KerasLayer has already been introduced in https://github.com/XanaduAI/pennylane/pull/529.

The tests take 7.45s to run locally - is this too much?

Closes https://github.com/XanaduAI/pennylane/issues/166.

### CODE REVIEW

File: pennylane/qnn/torch.py [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+
+                qnode = functools.partial(qnode, w)
+            else:
+                if self.input_is_default:  # The input argument can be positional or keyword
+                    qnode = functools.partial(qnode, **{self.input_arg: x})
+                else:
+                    qnode = functools.partial(qnode, x)
+        return qnode().type(x.dtype)
+
+    def __str__(self):
```
Comment (trbromley) [inline pennylane/qnn/torch.py:300]: Note that the `torch.nn` Module has an `extra_repr` method for adding more details like this, but it kept being shown in the methods section of the generated docs. So I chose this simpler option that we used for the KerasLayer.

File: pennylane/qnn/torch.py [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+        self.output_dim = output_dim[0] if isinstance(output_dim, Iterable) else output_dim
+
+        defaults = {
+            name for name, sig in self.sig.items() if sig.par.default != inspect.Parameter.empty
+        }
+        self.input_is_default = self.input_arg in defaults
+        if defaults - {self.input_arg} != set():
+            raise TypeError(
+                "Only the argument {} is permitted to have a default".format(self.input_arg)
+            )
```
Comment (trbromley) [inline pennylane/qnn/torch.py:249]: This bit is quite similar to the KerasLayer

File: tests/qnn/test_qnn_torch.py [hunk 3]
```diff
@@ -0,0 +1,420 @@
+# Copyright 2018-2020 Xanadu Quantum Technologies Inc.
```
Comment (trbromley) [inline tests/qnn/test_qnn_torch.py:1]: Note that the filename is chosen to avoid a repetition of `test_torch`.

File: tests/collections/conftest.py [hunk 4]
```diff
@@ -16,26 +16,6 @@
...
-        import tensorflow.contrib.eager as tfe
-
-        tf.enable_eager_execution()
-        Variable = tfe.Variable
-    else:
-        from tensorflow import Variable
-except ImportError as e:
-    tf = None
-    Variable = None
-
```
Comment (trbromley) [inline tests/collections/conftest.py:38]: The motivation for doing this is:
- A new `conftest` was introduced in the `tests/qnn` folder
- This resulted in failing tests
- The tests were failing because `tests/collections/test_collections.py` and `tests/collections/test_qnode_collection` include a `from conftest import torch, tf, Variable` line.
- Importing from conftest is not recommended because it can be ambiguous which one to import from (https://docs.pytest.org/en/2.7.3/plugins.html?highlight=re):
image
NOTE: This is from an older version of pytest, a more recent reference couldn't be found
- It made sense to make this change rather than simply removing the `tests/qnn/conftest.py` because importing conftest is not recommended.
- The result is a few more lines of code in `tests/collections/test_collections.py` and `tests/collections/test_qnode_collection`. Maybe we could do something smarter but it seems ok for now.

File: .travis.yml [hunk 5]
```diff
@@ -5,16 +5,29 @@ matrix:
   include:
   - python: 3.8
     dist: xenial
-    env: TF=0
```
Comment (trbromley) [inline .travis.yml:8]: The changes here are because I wanted to have a test where neither TF or Torch was installed.

**Comment by trbromley** [comment]
@josh146 I tagged you, was wondering if you'd have any idea why the test are failing :sweat_smile:

File: .travis.yml (hunk 5 above)
Comment (josh146) [reply .travis.yml:8]: I like this change, makes it easier to see which jobs use torch (it was too implicit before)

File: tests/collections/conftest.py (hunk 4 above)
Comment (josh146) [reply tests/collections/conftest.py:38]: @trbromley, the `__init__.py` didn't fix it? 🤔 I also wonder if `from .conftest import ...` would fix it.

**Comment by josh146** [comment]
@trbromley: looks like the only test that is failing is `TestCircuitDrawerIntegration`, and only on the Python 3.7 build. This sounds to me like a dependency issue; is that specific build using an older/newer version of numpy compared to the others?

File: tests/collections/conftest.py (hunk 4 above)
Comment (trbromley) [reply tests/collections/conftest.py:38]: Ah just revisited this and tried again:
- If we add `__init__.py` to `tests/collections` AND `tests/`, as well as update to `from .conftest`, the `test_collections`/`test_qnode_collection` will work
- However, the full test suite stops working: I get an error for the tests running `from gate_data import ...`
- This can be fixed by adding an `__init__.py` to all the required subfolders and changing to `from .gate_data import ...`
- I went a little bit down that route but there's so many files to change, it made me think that the solution currently adopted with code duplication is the lesser of two evils.

**Comment by trbromley** [comment]
> @trbromley: looks like the only test that is failing is `TestCircuitDrawerIntegration`, and only on the Python 3.7 build. This sounds to me like a dependency issue; is that specific build using an older/newer version of numpy compared to the others?

I changed the build to use Python 3.8 and it works. Not sure what the problem was :question:

File: tests/qnn/test_qnn_torch.py [hunk 6]
```diff
@@ -0,0 +1,421 @@
...
+@pytest.mark.usefixtures("get_circuit")
+class TestTorchLayer:
+    """Unit tests for the pennylane.qnn.torch.TorchLayer class."""
+
+    @pytest.mark.parametrize("n_qubits, output_dim", indices_up_to(1))
+    def test_no_torch(self, get_circuit, output_dim, monkeypatch):
+        """Test if an ImportError is raised when instantiated without PyTorch"""
+        c, w = get_circuit
+        with monkeypatch.context() as m:
+            m.setattr(qml.qnn.torch, "TORCH_IMPORTED", False)
```
Comment (josh146) [inline tests/qnn/test_qnn_torch.py:72]: oh nice! Better approach than monkeypatching `sys.modules`.

File: tests/qnn/test_qnn_torch.py [hunk 7]
```diff
@@ -0,0 +1,421 @@
...
+        clayer_weights = set("clayer{}.weight".format(i + 1) for i in range(3))
+        clayer_biases = set("clayer{}.bias".format(i + 1) for i in range(3))
+        qlayer_params = set(
+            "qlayer{}.w{}".format(i + 1, j + 1) for i in range(2) for j in range(len(w))
+        )
+
+        all_params = clayer_weights | clayer_biases | qlayer_params
+
+        assert dict_keys == all_params
+        assert len(dict_keys) == len(all_params)
```
Comment (josh146) [inline tests/qnn/test_qnn_torch.py:421]: Tests look good, but just wondering if there are any edge cases (weird output/input for QNodes) that we might be missing

File: tests/qnn/test_qnn_torch.py [hunk 8]
```diff
@@ -0,0 +1,421 @@
...
+
+
+def indices_up_to(n_max):
+    """Returns an iterator over the number of qubits and output dimension, up to value n_max.
+    The output dimension never exceeds the number of qubits."""
+    a, b = np.tril_indices(n_max)
+    return zip(*[a + 1, b + 1])
+
+
+@pytest.mark.usefixtures("get_circuit")
```
[the remark this answers is not in the document]
Comment (trbromley) [reply tests/qnn/test_qnn_torch.py:35]: Ok, done!

File: tests/qnn/test_qnn_torch.py (hunk 7 above)
Comment (trbromley) [reply tests/qnn/test_qnn_torch.py:421]: In terms of input I think we've specified quite well in the docstring:
image
I can think of edge cases where the second point is not satisfied, e.g., if someone puts a list of arrays as a single argument. We don't test for this case, do you think we need to? I think it's not needed since we say not to do this in the docstring and the user's `weight_shapes` dictionary also wouldn't have made sense in this case.

In terms of output, I think you pointed out one case where the QNode returns multiple arrays of different size. Is this something that's supported in PL generally, as I got an error as I mention in the comment thread above (https://github.com/XanaduAI/pennylane/pull/588/files#r420086869).

File: .travis.yml [hunk 9]
```diff
@@ -32,7 +45,7 @@ install:
 - pip install wheel pytest pytest-cov absl-py --upgrade
 - if [ "$TF" = "2" ]; then pip install tensorflow==2.1; fi
 - if [ "$TF" = "1" ]; then pip install tensorflow==1.13.2; fi
```
[the remark this answers is not in the document]
Comment (josh146) [reply .travis.yml:47]: Even if the existing 1.X code doesn't give us problems, I worry it will be confusing to users that some parts of the code base work with both, and some don't.

For the low-level non-Keras interface, it's quite easy to support both. We just need to make sure that conditionals or try-excepts are used to import eager-specific modules from the right place (import locations changes between 1.X and 2.X).

For the `qnn` module and Keras related code, however, I'm a lot more wary. Unless the `qnn` tests pass on TF 1.X, safer to assume it is only TF 2.0 compatible.

File: pennylane/qnn/torch.py [hunk 10]
```diff
@@ -0,0 +1,311 @@
...
+            function for initializing the QNode weights. If not specified, weights are randomly
+            initialized using the uniform distribution over :math:`[0, 2 \pi]`.
+
+    **Example**
+
+    .. code-block:: python
+
+        qlayer = qml.qnn.TorchLayer(qnode, weight_shapes, output_dim=2)
+        clayer = torch.nn.Linear(2, 2)
+        model = torch.nn.Sequential(qlayer, clayer)
```
[the remark this answers is not in the document]
Comment (josh146) [reply pennylane/qnn/torch.py]: Looks great!

File: pennylane/qnn/torch.py [hunk 11]
```diff
@@ -0,0 +1,311 @@
...
+        Args:
+            inputs (tensor): data to be processed
+
+        Returns:
+            tensor: output data
+        """
+        if len(inputs.shape) == 1:
+            return self._evaluate_qnode(inputs)
+
+        return torch.stack([self._evaluate_qnode(x) for x in inputs])
```
[the remark this answers is not in the document]
Comment (josh146) [reply pennylane/qnn/torch.py:273]: TypeError: can't convert np.ndarray of type numpy.object_. The only supported types are: float64, float32, float16, int64, int32, int16, int8, uint8, and bool.

^ Good to know, this is another 'gotcha' we should keep in mind in case of tech support inquiries. Namely, the Torch interface doesn't work for QNodes with 'ragged' output.

File: tests/qnn/test_qnn_torch.py (hunk 7 above)
Comment (josh146) [reply tests/qnn/test_qnn_torch.py:414]: Yep, I think the docstring does quite a good job!

File: pennylane/qnn/torch.py [hunk 12]
```diff
@@ -0,0 +1,309 @@
...
+            opt = torch.optim.SGD(model.parameters(), lr=0.5)
+            loss = torch.nn.L1Loss()
+
+        The model can be trained using:
+
+        .. code-block:: python
+
+            epochs = 8
+            batch_size = 5
+            batches = samples / batch_size
```
[the remark this answers is not in the document]
Comment (trbromley) [reply pennylane/qnn/torch.py:164]: Thanks @thisac, looks like a good improvement!

File: pennylane/qnn/keras.py [hunk 13]
```diff
@@ -186,7 +191,12 @@ def __init__(
         self, qnode, weight_shapes: dict, output_dim, weight_specs: Optional[dict] = None, **kwargs
     ):
         if not CORRECT_TF_VERSION:
-            raise ImportError("KerasLayer requires TensorFlow version 2 and above")
+            raise ImportError(
+                "KerasLayer requires TensorFlow version 2 and above. The latest "
```
[the remark this answers is not in the document]
Comment (trbromley) [reply pennylane/qnn/keras.py:195]: :thinking: I'm not sure, I think it depends on how you partition reading the sentence, but since you thought it looked better let's go with it

**Comment by josh146** [comment]
@mariaschuld this is out of scope for this PR, but I agree re: interfaces page. I think we should do the following:

* Add a short subsection on the Torch interface page https://pennylane.readthedocs.io/en/stable/introduction/interfaces/torch.html discussing `torch.nn`.

* Add cards on the top-level Interface quickstart linking directly to the `qml.qnn.TorchLayer` and `qml.qnn.KerasLayer` docstrings. The docstrings have very nice usage details, and probably no point duplicating it.

File: .travis.yml (hunk 9 above)
[the remark this answers is not in the document]
Comment (trbromley) [reply .travis.yml:47]: I just tried, and it looks nontrivial to get the Keras layers to work with 1.X.

I think we could get it to work but it would need some changes.

[the remark this answers is not in the document]
Comment (josh146) [reply .travis.yml:47]: Definitely not a priority, I think sunsetting 1.X support is much more likely than extending the qnn module to work with 1.X.

File: pennylane/qnn/torch.py [hunk 14]
```diff
@@ -0,0 +1,311 @@
...
+
+    Args:
+        qnode (qml.QNode): the PennyLane QNode to be converted into a Torch layer
+        weight_shapes (dict[str, tuple]): a dictionary mapping from all weights used in the QNode to
+            their corresponding shapes
+        output_dim (int): the output dimension of the QNode
+        init_method (callable): a `torch.nn.init <https://pytorch.org/docs/stable/nn.init.html>`__
+            function for initializing the QNode weights. If not specified, weights are randomly
+            initialized using the uniform distribution over :math:`[0, 2 \pi]`.
+
```
[the remark this answers is not in the document]
Comment (trbromley) [reply pennylane/qnn/torch.py:52]: Thanks for checking the docs building @mariaschuld!

I just rebuilt and had a look and it looks ok:
image

Note that there were a few intermediate changes since the version you built, not sure if that might have inadvertently fixed things :thinking:

File: pennylane/qnn/torch.py [hunk 15]
```diff
@@ -0,0 +1,311 @@
...
+        **Initializing weights**
+
+        The optional ``init_method`` argument of :class:`~.TorchLayer` allows for the initialization
+        method of the QNode weights to be specified. The function passed to the argument must be
+        from the `torch.nn.init <https://pytorch.org/docs/stable/nn.init.html>`__ module. For
+        example, weights can be randomly initialized from the normal distribution by passing:
+
+        .. code-block::
+
+            init_method = torch.nn.init.normal_
```
[the remark this answers is not in the document]
Comment (trbromley) [reply pennylane/qnn/torch.py:110]: That is weird :confused: 

It looks ok on the version I just built:
image

**Comment by trbromley** [comment]
@mariaschuld, I want to follow up to see if you still get those docs build errors? I couldn't replicate them when I tried to build locally so I'm wondering if it was either fixed or a system dependent problem?

Note that you requested changes so would be good to resolve. Thanks!

File: pennylane/qnn/torch.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (mariaschuld) [reply pennylane/qnn/torch.py:52]: Yes, it is working now :) But in the meantime I updated all my docs requirements, likely an old sphinx extension version or something...

File: pennylane/qnn/torch.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (mariaschuld) [reply pennylane/qnn/torch.py:110]: All good now !

File: .github/CHANGELOG.md [hunk 16]
```diff
@@ -1,6 +1,27 @@
 # Release 0.10.0 (development release)
 
 <h3>New features since last release</h3>
+* PennyLane QNodes can now be converted into Torch layers, allowing for creation of quantum and
```
Comment (josh146) [inline .github/CHANGELOG.md:4]:
```suggestion

* PennyLane QNodes can now be converted into Torch layers, allowing for creation of quantum and
```
